kio-gdrive: init at version 1.2.7

konstantinous opened the pull request:
I had to add signon, signon-ui, kaccounts-integration and providers
packages to be able to build kio-gdrive ( one of the packages from #62168 )

Package works well on my laptop. Signond daemon is probably not able to find it's plugins installed into user profile. However, it doesn't affect kio-gdrive itself.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Add google drive integration to KDE dolphin and other kio aware apps.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ttuegel commented on pkgs/applications/kde/kaccounts-integration.nix:
```diff
@@ -0,0 +1,45 @@
+
+    license = with licenses; [
+      gpl2
+      lgpl2.1
```
```suggestion
      lgpl21
```

ttuegel commented on pkgs/applications/kde/signon-kwallet-extension.nix:
```diff
@@ -0,0 +1,45 @@
+
+  preConfigure = ''
+    # mock signond plugins dir
+    sed 's/set(SIGNONEXTENSION_PLUGINDIR ''${_pkgconfig_invoke_result})/set(SIGNONEXTENSION_PLUGINDIR lib\/signon\/extensions )/g' -i cmake/modules/FindSignOnExtension.cmake
```
If you're replacing text verbatim (not using a regular expression) it's preferred to use [substituteInPlace](https://nixos.org/nixpkgs/manual/#ssec-stdenv-functions) because `sed` does not fail if there are no matches.

konstantinous replied:
Fixed

ttuegel commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13600,6 +13600,12 @@ in
 
   shhopt = callPackage ../development/libraries/shhopt { };
 
+  signond = libsForQt5.callPackage ../development/libraries/signond { };
```
If these derivations provide libraries, then they need to go into `mkLibsForQt5` instead of the top level.

konstantinous replied:
Fixed.

ttuegel commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19164,6 +19170,8 @@ in
 
   kid3 = libsForQt5.callPackage ../applications/audio/kid3 { };
 
+  kio-gdrive = kdeApplications.callPackage ../applications/misc/kio-gdrive { };
```
```suggestion
  kio-gdrive = libsForQt5.callPackage ../applications/misc/kio-gdrive { };
```

konstantinous replied:
Hello,
This package depends on some libraries from kde (e.g. libkgapi, kcalcore, kaccounts-integration). However, this package is not a part of "kde applications" set. I don't know what is the best way to provide these dependencies, but "libsForQt5.callPackage" will not work here...

ttuegel replied:
The KDE Applications libraries should be added to `mkLibsForQt5` [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L13321), it's probably just an oversight that they weren't already.

ttuegel reviewed (changes requested):
I have a few requests, but generally looks good! :smiley:

jonringer reviewed (commented):
signon-ui seems to fail to build

```
Project MESSAGE: ==== NOTE: Remember to add your API headers into `headers.files' for installation!
Project ERROR: accounts-qt5 development package not found
make[1]: *** [Makefile:44: sub-signon-ui-pro-make_first] Error 3
make[1]: Leaving directory '/build/signon-ui-3acb654/src'
make: *** [Makefile:92: sub-src-make_first-ordered] Error 2
builder for '/nix/store/cx6wbs1lrm47f4jzq09bgpj56xyqs0j7-signon-ui.drv' failed with exit code 2
error: build of '/nix/store/cx6wbs1lrm47f4jzq09bgpj56xyqs0j7-signon-ui.drv' failed
```

konstantinous commented:
accounts-qt5 is inside buildInputs, so the error is seems weird to me. And it works in my environment... Would you kindly give me more information so I could try to reproduce it?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-add-a-new-kde-package-for-kio-gdrive/5539/2

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -3837,6 +3837,11 @@
+    email = "konstantin@konstantsky.pw";
+    github = "konstantinous";
+    name = "Konstantin Khokhlov";
+  };
```
should be a separate commit:
```
maintainers: add konstantsky
```
also add githubId

konstantinous replied:
Done.

jonringer reviewed (commented):
each package addition should be it's own commit, and modules commits are generally labeled as:
```
nixos/module: concise message
```

Commit:
maintainers: add konstantsky

Commit:
signond: init at 8.60

Add signong binary in order to use online accounts integration in the future

Commit:
signon-ui: init at 0.17.02

Add signon-ui binary in order to use KDE online accounts integration in the future

Commit:
signon-plugin-oauth2: init at 0.24

Add oauth2 plugin for signond daemon to authenticate via google api

Commit:
kaccounts-integration: init

Add kaccounts-integration package from KDE applications in order to enable online accounts integration

Commit:
kaccounts-providers: init

Add kaccounts-providers from KDE applications in order to enable online accounts integration in the future

Commit:
accounts-qt: add graphviz to nativeBuildInputs as it is required by accounts-qt

Commit:
signon-kwallet-extension: init

Add signon-kwallet-extension from KDE applications in order to secure signon information

Commit:
ktp-accounts-kcm: init

Add ktp-accounts-kcm from KDE applications in order to enable KDE online accounts integration

Commit:
kio-gdrive: init at 1.2.7

Add google drive integration for KDE

Commit:
nixos/plasma5: add option to switch on/off online accounts support

It seems the only way to inject variables into user environment since
sidnon daemon requires a path to it's plugins and extensions. By default
it looks for plugins inside the directory in which it's binary is located

Commit:
accounts-qml-module: init at 0.7

Required by kde kcm kaccounts module

rnhmjoj reviewed (changes requested):
I'm not a KDE user, I could set up a test environment but I wouldn't know how to use this properly so I trust you saying it's working well.
On the packaging sides of things it looks good besides the bunch of small improvements that I mentioned. By the way, good job and thank you for contributing.

konstantinous commented on pkgs/development/libraries/accounts-qml-module/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  preConfigure = ''
+    export QT_PLUGIN_PATH="${qtbase.bin}/${qtbase.qtPluginPrefix}"
+    substituteInPlace src/src.pro --replace '$$[QT_INSTALL_BINS]' '${qtdeclarative.dev}/bin'
```
They are not needed anymore. You are right. Probably I forgot to pull latest changes before testing.

konstantinous commented on pkgs/development/libraries/accounts-qt/default.nix:
```diff
@@ -12,15 +12,20 @@ stdenv.mkDerivation rec {
-  preConfigure = ''
-    qmakeFlags="$qmakeFlags LIBDIR=$out/lib CMAKE_CONFIG_PATH=$out/lib/cmake"
-  '';
+#  preConfigure = ''
```
I removed commented lines created by me. Also I added a comment to explain why other lines were commented out. Who knows, maybe they are needed to build other packages.

rnhmjoj replied:
Ok, thank you.

konstantinous commented on nixos/modules/services/x11/desktop-managers/plasma5.nix:
```diff
@@ -158,6 +158,12 @@ in
         description = "Phonon audio backend to install.";
       };
+
+      enableOnlineAccounts = mkOption {
```
It's more convenient, thanks.

rnhmjoj commented on pkgs/applications/kde/signon-kwallet-extension.nix:
```diff
@@ -0,0 +1,45 @@
+
+  preConfigure = ''
+    # mock signond plugins dir
+    substituteInPlace cmake/modules/FindSignOnExtension.cmake --replace 'set(SIGNONEXTENSION_PLUGINDIR ''${_pkgconfig_invoke_result})' 'set(SIGNONEXTENSION_PLUGINDIR lib/signon/extensions)'
```
Thank you

rnhmjoj reviewed (approved):
Thank you. It looks good to me: I don't have any other comments to make.

rnhmjoj commented:
Ah, on last thing: before merging I would squash all the "refactor" commits.

konstantinous commented:
Hello @rnhmjoj , thank you for your advices. I hope I did exactly what you asked for.

rnhmjoj commented:
Not exactly, sorry. I meant moving the new changes to the respective `package: init at version`.
You can do this by doing

1. a soft reset to `HEAD~`, 
2. committing each change separately,
3. interactive rebase (`git rebase -i`) in which you reorder the commits and use the `fixup` action to merge the new change with the `package: init` commit.

The idea is that the history of a PR (like fixing mistakes) shouldn't be preserved when merged into master.

konstantinous commented:
Is there anything else I can do to help the merge happen?

rnhmjoj commented:
Sorry, I forgot about this PR: unfortunately github doesn't send notification on new commits.
The commits look good now.

rnhmjoj commented:
@jonringer can you finish your review?

jonringer commented on pkgs/development/libraries/signond/default.nix:
```diff
@@ -0,0 +1,58 @@
+
+  meta = with stdenv.lib; {
+    description = "A D-Bus service which performs user authentication on behalf of its clients.";
+    homepage = https://gitlab.com/accounts-sso/signond;
```
rfc 45
```suggestion
    homepage = "https://gitlab.com/accounts-sso/signond";
```

konstantinous replied:
Done.

jonringer commented on pkgs/development/libraries/signond/default.nix:
```diff
@@ -0,0 +1,58 @@
+    description = "A D-Bus service which performs user authentication on behalf of its clients.";
+    homepage = https://gitlab.com/accounts-sso/signond;
+    platforms = platforms.linux;
+    license = with licenses; [
+      lgpl21
+    ];
```
```suggestion
    license = with licenses; [ lgpl21 ];
```

konstantinous replied:
Fixed.

jonringer commented on pkgs/development/libraries/accounts-qml-module/default.nix:
```diff
@@ -0,0 +1,41 @@
+    mv src/Ubuntu  $out/${qtbase.qtQmlPrefix}
+  '';
+
+  outputs = [ "out" ];
```
Pretty sure this is default behavior
```suggestion
```

konstantinous replied:
Removed

jonringer commented on pkgs/development/libraries/accounts-qml-module/default.nix:
```diff
@@ -0,0 +1,41 @@
+{ stdenv, fetchFromGitLab, pkgconfig, qmake, accounts-qt, qttools, signond, qtdeclarative, qtbase }:
+
+stdenv.mkDerivation rec {
```
Please read https://nixos.org/nixpkgs/manual/#sec-language-qt on Qt packages in nix
```suggestion
mkDerivation rec {
```

konstantinous replied:
Done.

bachp commented:
@konstantinous Are you still working on this?  I would like to help you get this merged.

niknetniko commented:
> Hoever it needs a rebase and most probably an update of the mentioned components.

I believe in the kio-gdrive was added in #122040, so perhaps this is obsolete now.
